crates/radix-engine-toolkit/src/transaction_types/traverser/auxiliary/trusted_worktop/mod.rs

```diff
@@ -0,0 +1,338 @@
+
+    // Buckets tracking
+    bucket_tracker: BucketTracker,
+
+    // Worktop content tracking
+    worktop_content_tracker: WorktopContentTracker,
+}
+
+impl TrustedWorktop {
+    pub fn output(self) -> Vec<TrustedWorktopInstruction> {
+        self.trusted_state_per_instruction
+    }
+
+    fn add_new_instruction(
+        &mut self,
+        trusted: bool,
+        input_resources: Option<ResourceSpecifier>,
+    ) {
+        let resources = match input_resources {
+            Some(res) => {
+                if res.is_empty() {
+                    vec![]
+                } else {
+                    vec![res]
+                }
+            }
+            None => vec![],
+        };
+        self.trusted_state_per_instruction
+            .push(TrustedWorktopInstruction { trusted, resources });
+    }
+
+    fn add_new_instruction_from_bucket(&mut self, bucket: &Bucket) {
+        self.add_new_instruction(
+            bucket.is_known_resources(),
+            bucket.take_resources(),
+        );
+    }
+
+    fn add_new_instruction_with_many_resources(
+        &mut self,
+        trusted: bool,
+        resources: Vec<ResourceSpecifier>,
+    ) {
+        self.trusted_state_per_instruction
+            .push(TrustedWorktopInstruction { trusted, resources });
+    }
+}
+
+impl ManifestSummaryCallback for TrustedWorktop {
+    fn on_instruction(
+        &mut self,
+        instruction: &InstructionV1,
+        instruction_index: usize,
+    ) {
+        match instruction {
+            InstructionV1::TakeAllFromWorktop { resource_address } => {
+                if !self.worktop_content_tracker.is_untracked_mode() {
+                    if let Some(resources) = self
+                        .worktop_content_tracker
+                        .take_from_worktop_by_address(*resource_address)
+                    {
+                        self.bucket_tracker
+                            .new_bucket_known_resources(resources.clone());
+                        self.add_new_instruction(
+                            true,
+                            Some(resources.to_owned()),
+                        );
+                    } else {
+                        self.bucket_tracker.new_empty_bucket_known_resources();
+                        self.add_new_instruction(true, None)
+                    }
+                } else {
+                    // we don't know what is exactly on the worktop
+                    self.bucket_tracker.new_bucket_unknown_resources();
+                    self.add_new_instruction(false, None)
+                }
+            }
+            InstructionV1::TakeFromWorktop {
+                resource_address,
+                amount,
+            } => {
+                if !self.worktop_content_tracker.is_untracked_mode() {
+                    let resources =
+                        ResourceSpecifier::Amount(*resource_address, *amount);
+                    if amount.is_zero()
```

Comment:
I added this `amount.is_zero()` check because we can create empty bucket using `TakeFromWorktop` instruction at any time. There is no need to be any resource on the worktop in such a case, so worktop content tracker also doesn't have any stored resources of specified type. I've created dedicated test for that scenario: `trusted_worktop_take_zero()`.

Reply:
Also I fixed same case for Non Fungible resources and taking Non Fungible by amount == 0 (this instruction is trusted). There are 3 tests which verifies taking from worktop to empty buckets:
- `trusted_worktop_take_empty`
- `trusted_worktop_take_fungible_zero`
- `trusted_worktop_take_nonfungible_empty`